go-controller/pkg/node/healthcheck.go

```diff
@@ -59,54 +60,58 @@ func (l *loadBalancerHealthChecker) DeleteService(svc *kapi.Service) {
 
 func (l *loadBalancerHealthChecker) SyncServices(svcs []interface{}) {}
 
-func (l *loadBalancerHealthChecker) AddEndpoints(ep *kapi.Endpoints) {
-	name := ktypes.NamespacedName{Namespace: ep.Namespace, Name: ep.Name}
-	if _, exists := l.services[name]; exists {
-		l.endpoints[name] = countLocalEndpoints(ep, l.nodeName)
-		_ = l.server.SyncEndpoints(l.endpoints)
+func (l *loadBalancerHealthChecker) AddEndpointSlice(epSlice *discovery.EndpointSlice) {
+	svcName, ok := epSlice.Labels[discovery.LabelServiceName]
+	if ok && svcName != "" {
+		name := ktypes.NamespacedName{Namespace: epSlice.Namespace, Name: svcName}
+		if _, exists := l.services[name]; exists {
+			l.endpoints[name] = countReadyEndpoints(epSlice)
+			_ = l.server.SyncEndpoints(l.endpoints)
+		}
 	}
 }
 
-func (l *loadBalancerHealthChecker) UpdateEndpoints(old, new *kapi.Endpoints) {
-	name := ktypes.NamespacedName{Namespace: new.Namespace, Name: new.Name}
-	if _, exists := l.services[name]; exists {
-		l.endpoints[name] = countLocalEndpoints(new, l.nodeName)
-		_ = l.server.SyncEndpoints(l.endpoints)
+func (l *loadBalancerHealthChecker) UpdateEndpointSlice(oldEpSlice, newEpSlice *discovery.EndpointSlice) {
+	svcName, ok := newEpSlice.Labels[discovery.LabelServiceName]
+	if ok && svcName != "" {
+		name := ktypes.NamespacedName{Namespace: newEpSlice.Namespace, Name: svcName}
+		if _, exists := l.services[name]; exists {
+			l.endpoints[name] = countReadyEndpoints(newEpSlice)
+			_ = l.server.SyncEndpoints(l.endpoints)
+		}
 	}
-
 }
 
-func (l *loadBalancerHealthChecker) DeleteEndpoints(ep *kapi.Endpoints) {
-	name := ktypes.NamespacedName{Namespace: ep.Namespace, Name: ep.Name}
-	delete(l.endpoints, name)
-	_ = l.server.SyncEndpoints(l.endpoints)
+func (l *loadBalancerHealthChecker) DeleteEndpointSlice(epSlice *discovery.EndpointSlice) {
+	svcName, ok := epSlice.Labels[discovery.LabelServiceName]
+	if ok && svcName != "" {
+		name := ktypes.NamespacedName{Namespace: epSlice.Namespace, Name: svcName}
+		delete(l.endpoints, name)
+		_ = l.server.SyncEndpoints(l.endpoints)
+	}
 }
 
-func countLocalEndpoints(ep *kapi.Endpoints, nodeName string) int {
-	num := 0
-	for i := range ep.Subsets {
-		ss := &ep.Subsets[i]
-		for i := range ss.Addresses {
-			addr := &ss.Addresses[i]
-			if addr.NodeName != nil && *addr.NodeName == nodeName {
-				num++
-			}
+func countReadyEndpoints(epSlice *discovery.EndpointSlice) int {
+	var num int
+	for _, endpoint := range epSlice.Endpoints {
+		if endpoint.Conditions.Ready != nil && !*endpoint.Conditions.Ready {
```

Comment:
The [API states](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#endpointconditions-v1-discovery-k8s-io)

```
A nil value indicates an unknown state. In most cases consumers should interpret this unknown state as ready
``` 

But I guess it's better to be safe here and ensure we're actually getting ready endpoints